src/Foundatio.AWS/Queues/SQSQueue.cs

```diff
@@ -0,0 +1,269 @@
+            RegionEndpoint regionEndpoint = null,
+            TimeSpan? workItemTimeout = null,
+            TimeSpan? readQueueTimeout = null,
+            ISerializer serializer = null,
+            IEnumerable<IQueueBehavior<T>> behaviors = null,
+            ILoggerFactory loggerFactory = null)
+            : base(serializer, behaviors, loggerFactory) {
+
+            _queueName = queueName ?? throw new ArgumentNullException(nameof(queueName));
+
+            _credentials = credentials ?? FallbackCredentialsFactory.GetCredentials();
+            _regionEndpoint = regionEndpoint ?? RegionEndpoint.USEast1;
+            _workItemTimeout = workItemTimeout ?? TimeSpan.FromMinutes(5);
+            _readQueueTimeout = readQueueTimeout ?? TimeSpan.FromSeconds(2);
+
+            _client = new Lazy<AmazonSQSClient>(() => new AmazonSQSClient(_credentials, _regionEndpoint));
+        }
+
+        protected override async Task EnsureQueueCreatedAsync(CancellationToken cancellationToken = new CancellationToken()) {
+            if (!String.IsNullOrEmpty(_queueUrl))
+                return;
+
+            using (await _lock.LockAsync(cancellationToken).ConfigureAwait(false)) {
+                if (!String.IsNullOrEmpty(_queueUrl))
+                    return;
+
+                // per sqs documentation, create will get existing or create new queue
+                var createRequest = new CreateQueueRequest { QueueName = _queueName };
+                var createResponse = await _client.Value.CreateQueueAsync(createRequest, cancellationToken).ConfigureAwait(false);
+
+                _queueUrl = createResponse.QueueUrl;
+            }
+        }
+
+        protected override async Task<string> EnqueueImplAsync(T data) {
+            if (!await OnEnqueuingAsync(data).ConfigureAwait(false))
+                return null;
+
+            Interlocked.Increment(ref _enqueuedCount);
+            var message = new SendMessageRequest {
+                QueueUrl = _queueUrl,
+                MessageBody = await _serializer.SerializeToStringAsync(data).ConfigureAwait(false),
+            };
+
+            var response = await _client.Value.SendMessageAsync(message).ConfigureAwait(false);
+
+            var entry = new QueueEntry<T>(response.MessageId, data, this, SystemClock.UtcNow, 0);
+            await OnEnqueuedAsync(entry).ConfigureAwait(false);
+
+            return response.MessageId;
+        }
+
+        protected override async Task<IQueueEntry<T>> DequeueImplAsync(CancellationToken cancellationToken) {
+            var visibilityTimeout = Convert.ToInt32(_workItemTimeout.TotalSeconds);
+
+
+            // sqs doesn't support already canceled token, change timeout and token for sqs pattern
+            var waitTimeout = cancellationToken.IsCancellationRequested
+                ? 0
+                : Convert.ToInt32(_readQueueTimeout.TotalSeconds);
+
+            var cancel = cancellationToken.IsCancellationRequested
```

Comment:
In our other queue implementations (azure, redis and others) we use a linked cancellation token and async auto reset events (https://github.com/exceptionless/Foundatio/blob/master/src/Foundatio.AzureStorage/Queues/AzureStorageQueue.cs#L24) which we cancel in dispose.. We also use a do-while loop in at least 3 of our clients (pretty sure all off them do) 

https://github.com/exceptionless/Foundatio/blob/master/src/Foundatio.AzureStorage/Queues/AzureStorageQueue.cs#L93

https://github.com/exceptionless/Foundatio/blob/master/src/Foundatio.Redis/Queues/RedisQueue.cs#L264

You can run into some edge cases where you need to wait for a message to be received instead of returning immediately.  I guess if all tests pass it's good,